### PR TITLE
Refactor client to remove dependency on swizzling

### PR DESCRIPTION
**Pull request by cameroncooke**
## What's changed?

- Renamed `SBTUITunneledApplication` to `SBTUITestTunnelClient` and inherit from `NSObject` instead of `XCUIApplication`.
- Added `delegate` property to `SBTUITestTunnelClient`, delegates must conform to `SBTUITestTunnelClientDelegate`  protocol to hook into life-cycle processes.
- Created new sub-class of `XCUIApplication` named `SBTUITunneledApplication`.
- Both `SBTUITestTunnelClient` and `SBTUITunneledApplication` share the same interface apart form `launchWithOptions:statupBlock` which is now only implemented on `SBTUITunneledApplication`.
- `SBTUITunneledApplication` holds a strong reference to `SBTUITestTunnelClient` and proxies public API to the client instance.
- Added `ENABLE_UITUNNEL_SWIZZLING` pre-processor macro (defaults to `1`)  to enable swizzling to add `app` to `TestCase`, when `0` disables all swizzling.
- Added new `SBTUITestTunnel_TestsNoSwizzling` target which has the macro`ENABLE_UITUNNEL_SWIZZLING = 0` and uses a custom sub-class of `SBTUITunneledApplication` to customise behaviours to test that the non-swizzling technique works. Doesn't test all methods as they are implicitly tested in the swizzled tests as part of the `SBTUITestTunnel_Tests` target.

## Why?

Use of this library with our app which already makes heavy use of a custom `XCUIApplication` sub-class was near impossible due to the tight coupling of the client tunnelling logic and `XCUIApplication/XCTestCase`.

More specifically we already have complex life-cycle requirements in our tests that terminated and restarted the app as part of the test run and where the `launchArguments` were lost when re-launching due to the internal opaque `resetInternalState` method re-setting the state on termination.

The changes in this PR define a clear SOC between the client and `XCUIApplication/XCTestCase`. The new `SBTUITunneledApplication` class ensures that the public API is backwards compatible with existing code and disabling the swizzling behaviour is an opt-in. So no changes need to be made to existing code to use this implementation.

The requirement to inject an instance `XCUIApplication` in the initialiser of `SBTUITestTunnelClient` is deliberate in order to append the internal launch arguments i.e. `SBTUITunneledApplicationLaunchEnvironmentBonjourNameKey` etc.

The addition of a delegate defines a clear SOC and gives the developer the greatest level of flexibility though it's likely most of the time the delegate and injected application will be the same object. This is similar to how `UITableView().delegate` and `UITableView().datasource` which can be the same or separate objects.

## Non Swizzled Examples

To use without swizzling set the pre-processor macro `ENABLE_UITUNNEL_SWIZZLING ` to `0` and the create your own instance of `XCUIApplication`.

Sub-classing `SBTUITunneledApplication`:

```swift
class MyCustomApplication: SBTUITunneledApplication {
    // custom methods and behaviours
}

class MyTests: XCTestCase {

    let app = MyCustomApplication()

    override func setUp() {
        super.setUp()

        app.launchTunnel()
    }

    func testSomething() {....}
}
```

Sub-classing `XCUIApplication` directly:

```swift
class MyCustomApplication: XCUIApplication {
    private lazy var client: SBTUITestTunnelClient = {
        let client = SBTUITestTunnelClient(application: self)!
        client.delegate = self
        return client
    }()

    func launchApp() {
        // do any custom luanch things
        client.launchTunnel()
    }
}

extension MyCustomApplication: SBTUITestTunnelClientProtocol {
    // Add methods that proxy to client 
    // or just make `client` public.
}

extension MyCustomApplication: SBTUITestTunnelClientDelegate {
    func testTunnelClientIsReady(toLaunch sender: SBTUITestTunnelClient) {
        launch()
    }

    func testTunnelClient(_ sender: SBTUITestTunnelClient, didShutdownWithError error: Error?) {
        // handle errors
        terminate()
    }
}

class MyTests: XCTestCase {

    let app = MyCustomApplication()

    override func setUp() {
        super.setUp()

        app.launchTunnel()
    }
  
    func testSomething() {....}
}
```

Concrete example here: [Example/SBTUITestTunnel_TestsNoSwizzling/NoSwizzlingTests.swift](https://github.com/Subito-it/SBTUITestTunnel/blob/bcd6dcfa238650dbbd784e121de87324a2bc7caf/Example/SBTUITestTunnel_TestsNoSwizzling/NoSwizzlingTests.swift#L1)

### CODE REVIEW

File: Pod/Client/SBTUITunneledApplication.h [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// SBTUITunneledApplication.h
+// SBTUITestTunnelClient.h
```
Comment (tcamin) [inline Pod/Client/SBTUITunneledApplication.h]: Comment doesn't match filename

File: Pod/Client/SBTUITunneledApplication.m [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// SBTUITunneledApplication.m
+// SBTUITestTunnelClient.m
```
Comment (tcamin) [inline Pod/Client/SBTUITunneledApplication.m]: Comment doesn't match filename

File: Pod/Client/SBTUITunneledApplication.m [hunk 3]
```diff
@@ -15,764 +15,347 @@
...
-    NSLog(@"[SBTUITestTunnel] Resolving bonjour service %@", self.bonjourName);
-    [self.bonjourBrowser resolveWithTimeout:self.connectionTimeout];
-    
     [self launch];
-    
-    [self waitForAppReady];
 }
 
-- (void)launchConnectionless:(NSString * (^)(NSString *, NSDictionary<NSString *, NSString *> *))command
+- (void)testTunnelClient:(nonnull SBTUITestTunnelClient *)sender didShutdownWithError:(NSError * _Nullable)error
```
Comment (tcamin) [inline Pod/Client/SBTUITunneledApplication.m]: Given that nullability attributes are not required in implementation files I'd rather remove them altogether.

File: Pod/Client/SBTUITestTunnelClient.h [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+ *  The delegate must adopt the SBTUITunneledApplicationDelegate protocol. The delegate is not retained.
+ */
+@property (nonatomic, weak) id<SBTUITestTunnelClientDelegate> _Nullable delegate;
+
+/**
+ *  Create an instance of SBTUITestTunnelClient passing in an XCUIApplication to use.
+ *
+ *  @param application The instance of XCUIApplication.
+ */
+- (nullable instancetype)initWithApplication:(nonnull XCUIApplication *)application;
```
[the remark this answers is not in the document]
Comment (ryan-noths) [reply Pod/Client/SBTUITestTunnelClient.h]: Updated in most recent commit

File: Pod/Client/SBTUITestTunnelClient.m [hunk 5]
```diff
@@ -0,0 +1,789 @@
...
+@property (nonatomic, copy) NSDictionary <NSString *, NSString *> *initialLaunchEnvironment;
+@property (nonatomic, strong) NSString *(^connectionlessBlock)(NSString *, NSDictionary<NSString *, NSString *> *);
+
+@end
+
+@implementation SBTUITestTunnelClient
+
+static NSTimeInterval SBTUITunneledApplicationDefaultTimeout = 30.0;
+
+- (nullable instancetype)initWithApplication:(nonnull XCUIApplication *)application
```
[the remark this answers is not in the document]
Comment (ryan-noths) [reply Pod/Client/SBTUITestTunnelClient.m]: Removed all nullability attributes from implementation files in most recent commit

File: Pod/Client/SBTUITunneledApplication.m (hunk 3 above)
Comment (ryan-noths) [reply Pod/Client/SBTUITunneledApplication.m]: Removed all nullability attributes from implementation files in most recent commit

File: Pod/Client/SBTUITunneledApplication.h (hunk 1 above)
Comment (ryan-noths) [reply Pod/Client/SBTUITunneledApplication.h]: Updated 👍

File: Pod/Client/SBTUITunneledApplication.m (hunk 2 above)
Comment (ryan-noths) [reply Pod/Client/SBTUITunneledApplication.m]: Updated 👍

File: Pod/Client/SBTUITestTunnelClient.h [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ *  Create an instance of SBTUITestTunnelClientDelegate.
+ *
+ *  The methods adopted by the object you use to manage the application life-cycle usually an instance of XCUIApplication.
+ */
+@protocol SBTUITestTunnelClientDelegate <NSObject>
+
+/**
+ Informs the delegate that it should launch the XCUIApplication under-test before the tunnel is established.
+ It's required that you avoid lunching until the delegate is called.
```
[the remark this answers is not in the document]
Comment (ryan-noths) [reply Pod/Client/SBTUITestTunnelClient.h]: Fixed in latest commit

**Comment by ryan-noths** [comment]
Appears some unit tests are broken now, I'm taking a look now and I'll get back to you when it's ready

**Comment by ryan-noths** [comment]
All seems good now 👍

**Comment by tcamin** [comment]
Awesome! Before merging could you also update the documentation to reflect your changes?

**Comment by cameroncooke** [comment]
So we just doing some final testing with our use-cases and then if all good we'll update the README.

**Comment by cameroncooke** [comment]
@tcamin Just added documentation, so hopefully we're good to go?

**Pull request merged**